File: STRIVER_SDE_SHEET/Arrays_and_Hashing/FourSum.py
n = int(input())
arr = [0]*n
for i in range(n):
    arr[i] = int(input())

trgt = int(input())

ans = list()
arr.sort()
for i in range(n):
    if i>0 and arr[i]==arr[i-1]: continue
    for j in range(i+1,n):
        if(j>i+1 and arr[j]==arr[j-1]): continue
        k=j+1
        l=n-1
        while(k<l):
            sum = arr[i]+arr[j]+arr[k]+arr[l]
            if(sum<trgt):
                    k+=1
            elif sum>trgt:
                    l-=1
            else:
                temp = [arr[i],arr[j],arr[k],arr[l]]
                temp.sort()
                ans.append(list(temp))
                k+=1
                l-=1
                while k<l and arr[k]==arr[k-1]: k+=1
                while k<l and arr[l]==arr[l+1]: l-=1

for s in ans:
    print("[",s[0],s[1],s[2],s[3],"]")

                
                
        

File: Arrays_Series_Striver/FourSum.java
package Arrays_Series_Striver;
import java.util.*;
public class FourSum {
        public List<List<Integer>> fourSum(int[] arr, int trgt) {
        List<List<Integer>> ans = new ArrayList<>();
        Arrays.sort(arr);
        int n = arr.length;
        for(int a=0;a<n;a++){
            if(a>0&&arr[a-1]==arr[a]) continue;
            for(int b=a+1;b<n;b++){
                if(b>a+1&&arr[b-1]==arr[b]) continue;
                int c = b+1;
                int d = n-1;
                while(c<d){
                    long sum = 0L+ arr[a]+arr[b]+arr[c]+arr[d];
                    if(sum<trgt) c++;
                    else if(sum>trgt) d--;
                    else{
                        ans.add(Arrays.asList(arr[a],arr[b],arr[c],arr[d]));
                        while(c<d&&arr[c+1]==arr[c]) c++;
                        while(c<d&&arr[d-1]==arr[d]) d--;
                        c++;
                        d--;
                    }
                }
            }
        }
        return ans;
    }
}
